qualtran/bloqs/mcmt/specialized_ctrl.py

```diff
@@ -0,0 +1,255 @@
+
+            out_soqs = bb.add_d(ctrl_bloq, **in_soqs)
+
+            ctrl = out_soqs.pop(ctrl_reg_name)
+            return [ctrl], out_soqs.values()
+
+    else:
+        # the difficult case: must combine the two controls into one
+        ctrl_1_bloq_and_reg_name = get_ctrl_bloq_and_ctrl_reg_name(1)
+        assert (
+            ctrl_1_bloq_and_reg_name is not None
+        ), "invalid usage: controlled-by-1 variant must be provided"
+        ctrl_1_bloq, ctrl_reg_name = ctrl_1_bloq_and_reg_name
+
+        ctrl_bloq = _MultiControlledFromSinglyControlled(
+            cvs=(ctrl_bit, current_ctrl_bit), ctrl_bloq=ctrl_1_bloq, ctrl_reg_name=ctrl_reg_name
+        )
+
+        def _adder(
+            bb: 'BloqBuilder', ctrl_soqs: Sequence['SoquetT'], in_soqs: dict[str, 'SoquetT']
+        ) -> tuple[Iterable['SoquetT'], Iterable['SoquetT']]:
+            # extract the two control bits
+            (ctrl0,) = ctrl_soqs
+            ctrl1 = in_soqs.pop(ctrl_reg_name)
+
+            ctrl0 = cast(Soquet, ctrl0)
+            ctrl1 = cast(Soquet, ctrl1)
+
+            # add the singly controlled bloq
+            in_soqs |= {ctrl_reg_name: np.array([ctrl0, ctrl1])}
+            ctrls, *out_soqs = bb.add_t(ctrl_bloq, **in_soqs)
+            assert isinstance(ctrls, np.ndarray)
+            ctrl0, ctrl1 = ctrls
+
+            return [ctrl0], [ctrl1, *out_soqs]
+
+    return ctrl_bloq, _adder
+
+
+def get_ctrl_system_1bit_cv(
+    bloq: 'Bloq',
+    ctrl_spec: 'CtrlSpec',
+    *,
+    current_ctrl_bit: Optional['ControlBit'],
+    get_ctrl_bloq_and_ctrl_reg_name: Callable[['ControlBit'], tuple['Bloq', str]],
+) -> tuple['Bloq', 'AddControlledT']:
+    """Build the control system for a bloq with specialized single-qubit controlled variants.
+
+    Uses the provided specialized implementation when a singly-controlled variant of the bloq is
+    requested. When controlled by multiple qubits, the controls are reduced to a single qubit
+    and the singly-controlled bloq is used.
+
+    The user must provide two specializations for the bloq: controlled by `1` and by `0`.
+
+    When only one specialization (controlled by `1`) is known, use
+    :meth:`get_ctrl_system_1bit_cv_from_bloqs` instead.
+
+    Args:
+        bloq: The current bloq.
+        ctrl_spec: The control specification
+        current_ctrl_bit: The control bit of the current bloq, one of `0, 1, None`.
+        get_ctrl_bloq_and_ctrl_reg_name: A callable that accepts a control bit (`0` or `1`),
+            and returns the controlled variant of this bloq and the name of the control register.
+    """
+    return _get_ctrl_system_1bit_cv(
+        bloq,
+        ctrl_spec,
+        current_ctrl_bit=current_ctrl_bit,
+        get_ctrl_bloq_and_ctrl_reg_name=get_ctrl_bloq_and_ctrl_reg_name,
+    )
+
+
+def get_ctrl_system_1bit_cv_from_bloqs(
+    bloq: 'Bloq',
+    ctrl_spec: 'CtrlSpec',
+    *,
+    current_ctrl_bit: Optional['ControlBit'],
+    bloq_with_ctrl: 'Bloq',
```

Comment:
is this now just `ctrl_bloq`? I think the "with" phrasing was to support passing in both positive and negative control

Reply:
yes, this is the controlled variant of the bloq. I wanted to not confuse it with `bloq.controlled()` (i.e. when `current_ctrl_bit` is `1`, `bloq_with_ctrl` is `bloq`), so was hesitant to name it `ctrl_bloq`. But I think `ctrl_bloq` is also okay.

Reply:
yeah that might be worth keeping distinct